### ISSUE TITLE
Parameters<typeof someFunction> does not preserve type constraint or parameter inference.

### ISSUE THREAD

**Issue by DrSammyD** [microsoft/TypeScript#54014]
Given the following two types.

```ts
type Channel = {
  [key: string]: number;
};

type T<C extends Channel> = {
  channel: C;
  props: (keyof C)[];
};
```

The following function is able to correctly constrain the type of the second parameter given the following function declaration and call.

```ts
function createObject<C extends Channel>(channel: C, props: (keyof C)[]): T<C> {
  return {
    channel,
    props,
  };
}

const someObject = createObject(
  {
    Stuff: 1,
    Yarn: 2,
  },
  ['Stuff', 'Yarn', 'Thing'] // If you try to add a string that is not a key of 'channel', you'll get a type error. Thing is inferred invalid, and I don't have to specify a generic property
);
```

However, this constraint is not preserved in the `Parameters` utility type.
```ts
let thing: Parameters<typeof createObject> = [
  {
    "Stuff": 1,
    "Yarn": 2,
  },
  ["Yarn", "Stuff", "Thing"] // So... that constraint wasn't preserved.
]
```

At least not until the generic is explicitly set.

```ts
let thingAgain: Parameters<typeof createObject<{"Stuff":number;"Yarn":number}>> = [
  {
    "Stuff": 1,
    "Yarn": 2,
  },
  ["Yarn", "Stuff", "Thing"] // And we're back!
]
```

Ultimately, the desired functionality is to have types obtain the same type inference as functions without a runtime requirement to keep the typing/code dry.

[Playground link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAwgFgQwHZIgGygXigbwFBRQDaA1hCAFxQDOwATgJZIDmAulUgK4C2ARhHQDceAL7C8oSFAAqAHhhQIAD2AQkAE2qxEKdAD4suAlADGO1GioxhhMHQD2YalQAUZEPYBmsAJRFWwmJ4eJ6cSCbADPZIpnQQCKoA8rwAVhAR8ooqaprayBZ6Lmb56FYANFB2js5QbuRevv4+VHIwBviEccCcdDEdhKbm6GXGtg5OI4RBIsEm0bQ09twQyWkRhiZxCSup6cAuxv2EAMrdnp5UAIyTAwCaCL1UAEw3IjdEAOSnnOcfFR-3Xp-KAfaRwJjMD6sKAAehhUAAkt4PJwoPQQGj7FAEOp1NiaPQIWjEMAoAwtEh7KSEFB3FAGh9iro0MCUR80BhmBBqWjwNABA46AA6GTglhkrRMTwCiB4pgANwQaAY6gqyDxCKg6miH1JiHl0GAWOokBMDE8GJpXNQjBMlXGAlAeB84jQ3OJEKoAAUHghlqo6NRZJIIA1NvEkrsIgZsERDqMoAAib7nRNXG6EROApBpqAvYxvYxELMPHMVZNnTyJ8tgiGJ6FwqDHexC1vEhKmeb0BBMUkAdwQ1CQuvtEGoAgN6iFeFYwTdpOAYuYAEFmD2kN7ff6BEGQ2GtpG1sBZDgKz8qxQuHwBIIS7001f+HQRHoY8R4wMkymL1Brgm7zmzyvO8AHVl+lZgYmtYsPWsLwsuGhQH2EAfHEUC8AgJgkAAhDOQA)

**Comment by fatcerberus** [comment]
This would first require implementing a new kind of a generic type that would have its type parameter(s) inferred on assignment, as well as the ability for type aliases to directly accept and return generic types (HKT).  Neither of these are possible with the current type system.

**Comment by fatcerberus** [comment]
> a new kind of a generic type that would have its type parameter(s) inferred on assignment

For this, see https://github.com/microsoft/TypeScript/issues/47755.

> the ability for type aliases to directly accept and return generic types (HKT)

For HKT, there are several issues, notably #45438, #44875, and the seminal issue #1213.

**Comment by DrSammyD** [comment]
Thanks @fatcerberus for the other issues.

**Comment by DrSammyD** [comment]
I'm going to reopen this, because even if those other suggestions are implemented, Parameters will probably still be unconstrained.

**Comment by RyanCavanaugh** [comment]
To expedite the triage process, we need everyone to follow the issue template and instructions.

When you clicked "Create New Issue", the issue form was pre-populated with a template and some instructions. We need you to read those instructions *completely*, follow them, and then fill in *all* the fields in that template.

We are not able to assist with issues that don't follow the template instructions as they represent a significantly larger amount of work compared to issues which are correctly specified. Thank you for understanding.
